### PR TITLE
Add RedundantHigherOrderMapUsage rule

### LINKED ISSUE #2975: Map { it } must return an error
```
        val data = listOf(0, 1, 2)

        val output = data.map { it }
```

## Expected Behavior of the rule
Detekt return an error. For example map operator don't map

Are Detekt has rule for the case?

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #2975

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsageSpec.kt [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                            it 
+                        }
+                        .filter { it > 1 }
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings).hasSourceLocation(5, 10)
+            assertThat(findings[0]).hasMessage("This 'map' call can be replaced with 'onEach' or 'forEach'.")
+        }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsageSpec.kt:47]: This! 👏👏👏👏👏 I saw it SO MANY TIMES! A map that should be replaced with an `onEach`. And it slow down the reading of the code a lot.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsage.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import org.jetbrains.kotlin.psi.unpackFunctionLiteral
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.bindingContextUtil.getTargetFunctionDescriptor
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.types.KotlinType
+
+/**
+ * Turn on this rule to flag redundant 'map' calls.
```
Comment (BraisGabin) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsage.kt:28]: I would change the description with something like:

Redundant maps add complexity to the code and accomplish nothing. They should be removed or replaced with the proper operator.

**Review by BraisGabin — APPROVED** [review]
I really like this rule. The tests are really good and you are checking a lot of edge cases. Great work 👏.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsage.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        val message = if (lambdaStatements.size == 1) {
+            if (receiverIsSet) {
+                "This 'map' call can be replaced with 'toList'."
+            } else {
+                "This 'map' call can be removed."
+            }
+        } else {
+            "This 'map' call can be replaced with 'onEach' or 'forEach'."
+        }
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantHigherOrderMapUsage.kt:108]: Can you convert this to a `when`?

**Review by cortinico — APPROVED** [review]
Great addition 👍

**Review by cortinico — APPROVED** [review]

**Review by BraisGabin — APPROVED** [review]

**Review by schalkms — APPROVED** [review]
